ToggleGroupControl: Improve styling for icon options

mirka opened the pull request:
Part of #36735
Follow-up to https://github.com/WordPress/gutenberg/pull/39760#issuecomment-1081861004

## What?

- Tweak the ToggleGroupControlOptionIcon styles so they have [square proportions](https://github.com/WordPress/gutenberg/pull/39760#issuecomment-1081861004).
- Add an `__experimentalIsIconGroup ` variant to ToggleGroupControl, recommended for use with icon options.
- Update the ToggleGroupControlOptionIcon component to always show a tooltip and to require a label prop for improved accessibility

## Why?

In preparation for migrating TextTransformControl/TextDecorationControl to a ToggleGroupControl implementation so the semantics are correct, and to get a large size variant "for free".

## How?

ToggleGroupControlOptionIcon needs to be a different size based on the size variant of the ToggleGroupControl, so we get that via the Context System.

## Testing Instructions

1. `npm run storybook:dev` and see the ToggleGroupControl With Icons story.

## Screenshots or screencast <!-- if applicable -->

Note that the **focus state** shows the border surrounding the entire ToggleGroupControl, not the individual option. I think this is a good way to maintain consistency within the ToggleGroupControl, and signal that the options are mutually exclusive.

In contrast, if we eventually change the TextDecorationControl implementation to be non-mutually-exclusive (probably semantically a group of checkboxes), the focus ring for that component could surround each individual option.

The is a subtle difference but an important affordance, because the standard focus ring moves with the <kbd>Tab</kbd> key, while radio options move with the arrow keys.

https://user-images.githubusercontent.com/555336/183407354-e591959b-5d0f-4a5f-be7d-995ef504f656.mp4

mirka committed:
ToggleGroupControl: Add `isBorderless` variant

mirka committed:
ToggleGroupControlOptionIcon: Improve sizing

mirka committed:
Add changelog

mirka committed:
Update snapshots

jasmussen commented:
This is sweet! 

Here's before:
![before](https://user-images.githubusercontent.com/1204802/183584946-7f2f172f-103b-4266-90e6-53a6f390ba04.gif)

Here's storybook:

![storybook](https://user-images.githubusercontent.com/1204802/183585141-63f07010-733b-4955-93f2-117368c6d055.gif)

I think the border-less and with-icons version of the segmented control will be useful, the focus style makes sense (since arrow keys are meant to pick the choice), and it even makes sense to replace the text decoration control since all those options are mutually exclusive.

One thing that jumps out to me is that the toggled state is slightly rectangular:
<img width="397" alt="Screenshot 2022-08-09 at 09 05 28" src="https://user-images.githubusercontent.com/1204802/183585874-9b3a163a-2fb0-40f5-9a43-a85775cd27be.png">

That appears to be because the height of the toggled state inside the current segmented control is also 30px (36px current size minus top/bottom borders, minus 2px inner padding top and bottom = 30px). Ideally the toggled state remains square, which would make it 30x30, but perhaps because this variant is borderless, we can keep the state 32x32?

Perhaps not something to block this PR, since we should be optimizing for the future 40px size, which would make the toggled states inside 34x34:

<img width="273" alt="Screenshot 2022-08-09 at 09 08 23" src="https://user-images.githubusercontent.com/1204802/183586435-5b15ddeb-ac30-4314-9342-1674ccf03d7d.png">

But even in that case, it would be nice for the toggled states of the iconless version to be the full 40x40 sizes. Like so:
<img width="163" alt="Screenshot 2022-08-09 at 09 08 27" src="https://user-images.githubusercontent.com/1204802/183586497-5bd862b3-ff89-4544-82b7-f08c85aab359.png">

(Speaking of, that group transform interface would also benefit from this control.)

ciampo commented on packages/components/src/toggle-group-control/toggle-group-control/styles.ts:
```diff
@@ -28,18 +27,22 @@ export const ToggleGroupControl = ( {
-	}
-
-	&:focus-within {
+		&:focus-within {
```
Nit: indentation

```suggestion
	&:focus-within {
```

mirka replied:
Oh good catch, thanks

mirka committed:
Fix square size

ciampo reviewed (commented):
Code changes look good!

I'm only a bit reluctant on adding a prop that is explicitly tied to the visual aspect of the component, which may change over time and make that prop obsolete (i.e. we may decide to always show borders, or to never show borders).

Also, a related question: have we considered "unifying" the look of our toggle controls (aka segmented controls) and always (or never) show borders?

mirka commented:
> the toggled state is slightly rectangular

My bad, fixed in fa743f7 ✨ 

> toggled states of the iconless version to be the full 40x40 sizes

I definitely see your point. Since this will require quite a bit more reworking of the code though, I'm hoping we can first play around with it in context when I post a PR to put this into the actual Typography panel. It might be that the "mis-sized" button looks weird, but it also might be that a "mis-sized" focus ring looks weird. That should help us assess how urgent it is to change that detail.

mirka committed:
Fix indentation

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

mirka commented:
> I'm only a bit reluctant on adding a prop that is explicitly tied to the visual aspect of the component, which may change over time and make that prop obsolete (i.e. we may decide to always show borders, or to never show borders).

Yes, I hear you. Would you prefer we name it "isIconGroup" or something like that?

@jasmussen Can we pretty much assume that the borderless style will only be for icon button groups? It will also inform how we phrase the style guidance:

https://github.com/WordPress/gutenberg/blob/901814ef6d5b9c0de080a82957a4cd8b312ee1a7/packages/components/src/toggle-group-control/stories/index.js#L142-L145

ciampo commented on packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx:
```diff
@@ -60,8 +60,10 @@ function ToggleGroupControlOptionBase(
 	const {
 		className,
 		isBlock = false,
+		isIcon = false,
 		value,
 		children,
+		size = 'default',
```
It's weird that TypeScript is allowing this — technically `isIcon` and `size` are not defined on neither `toggleGroupControlContext` nor `buttonProps`.

I'm not sure if we can pass props that are not part of `ToggleGroupControlOptionBase`'s props via the "context system". If we want to pass "private", internal props, we may need to use `useToggleGroupControlContext` (and declare these props in the `ToggleGroupControlContextProps` type)

mirka replied:
> It's weird that TypeScript is allowing this

I don't quite understand it either, good catch.

I fixed this up. `size` is now private to the context system, and `isIcon` is defined on OptionBase but not the final Option components. Final props tables of the Option components can now be verified in Storybook.

ciampo reviewed (commented):
> Yes, I hear you. Would you prefer we name it "isIconGroup" or something like that?

What about a `variant` prop that accepts an enum of possible values, i.e. `default` and `minimal` (for the borderless version) ? Or any other better name.

jasmussen commented:
> Can we pretty much assume that the borderless style will only be for icon button groups? It will also inform how we phrase the style guidance:

Yes, I'd think that would be a safe bet.

mirka commented:
> What about a `variant` prop that accepts an enum of possible values, i.e. `default` and `minimal` (for the borderless version) ? Or any other better name.

Given Joen's guidance, I'm now inclined to introduce a new facade component `ToggleIconGroupControl` instead of trying to do this as a variant of `ToggleGroupControl`. I think this would make it simpler in terms of ergonomics, discoverability, documentation, etc. I'll start exploring this, let me know if you'd prefer to stay with the `variant` approach @ciampo.

mirka committed:
Fix a11y labeling

ciampo commented:
Exploring on a potential `ToggleIconGroupControl` component sounds good to me 👍

jasmussen commented:
I don't necessarily think it's the most urgent thing, I trust you can balance it against other tasks on deck, I know there's a lot.

mirka committed:
Fix types

mirka committed:
Rename `isBorderless` prop

mirka committed:
Fixup

mirka committed:
Update snapshots

mirka committed:
Fix up story control

mirka committed:
Privatize `size` context

mirka commented:
@ciampo The facade component refactor is a bit too big for the scope of this PR, so if it's ok with you I'd like to merge this PR in the current state and move the refactor to a follow-up. I renamed `isBorderless` to `__experimentalIsIconGroup` for the time being, and added a TODO comment.

mirka committed:
Fixup docs

ciampo commented on packages/components/src/toggle-group-control/stories/index.js:
```diff
@@ -151,20 +157,19 @@ export const WithIcons = ( props ) => {
 				icon={ formatLowercase }
-				showTooltip={ true }
-				aria-label="Lowercase"
+				label="Lowercase"
```
We should update also the [unit tests](https://github.com/WordPress/gutenberg/blob/98a478f05bbd645e15d2b4b23525c7a00e0ba431/packages/components/src/toggle-group-control/test/index.js#L59-L70) and probably add a changelog entry about adding the `label` prop to the component and always showing a tooltip?

mirka replied:
Great idea. I've also converted the test file to a `tsx`. (no other errors! yay)

ciampo reviewed (approved):
Changes seem to test well on Storybook 🚀 

It's not super easy to check for regressions re. size, label and tooltip behavior on the remaining Storybook stories because of the mix of knobs and controls (also, since the stories are still written in JavaScript, we don't get the benefits of the auto-updating controls that we'd otherwise have).

mirka committed:
Update unit tests

mirka committed:
Add changelog about `label` prop

mirka committed:
Merge branch 'trunk' into toggle-group-control-icon-styles

mirka commented:
> It's not super easy to check for regressions re. size, label and tooltip behavior on the remaining Storybook stories because of the mix of knobs and controls

Yeah, I've been putting off the Storybook cleanup for a while 😇 I'll do it before the refactor.

mirka committed:
Rename snapshot
